src/check_result.jl

```diff
@@ -30,8 +30,8 @@ for (T1, T2) in ((AbstractThunk, Any), (AbstractThunk, AbstractThunk), (Any, Abs
     end
 end
 
-test_approx(::ZeroTangent, x, msg=""; kwargs...) = test_approx(zero(x), x, msg; kwargs...)
-test_approx(x, ::ZeroTangent, msg=""; kwargs...) = test_approx(x, zero(x), msg; kwargs...)
+test_approx(::AbstractZero, x, msg=""; kwargs...) = test_approx(zero(x), x, msg; kwargs...)
+test_approx(x, ::AbstractZero, msg=""; kwargs...) = test_approx(x, zero(x), msg; kwargs...)
```

Comment:
The reason I've changed this is somewhat convoluted and the dependency goes the wrong way. But hear me out:

When writing a helper `zygote_rrule` function in Zygote (see https://github.com/FluxML/Zygote.jl/pull/987/) one needs to transform `nothing` (representing a zero gradient) to either `ZeroTangent` or `NoTangent`.
If we decide to transform to `ZeroTangent`, `test_rrule` complains that `NoTangent` should be used for non-perturbable arguments (say indices).
If we decide to transform to `NoTangent`, the `test_approx` function throws a fit, since `test_approx(::NoTangent, x)` is not defined and it falls back to the default implementation.

Alternatives:
- Do not provide `zygote_rrule` function in Zygote. It would still be possible to write a conversion for each separate function (where we know which `AbstractZero` the `nothing` represents). This seems to be the cleanest option but on the other hand it would greatly reduce the usability of `rrule_f` kwarg for testing Zygote gradients, since one would need to write a custom rrule wrapper for each function that is being tested. Eww.
- Pirate `test_approx(::NoTangent, x)` in Zygote. (probably very bad)

All of the above aside, I am leaning towards the current state where `test_approx(::NoTangent, x)` is not defined. However in this case I suspect it would be better to be practical. After all, we are happy to define `+(NoTangent, x)` to be equal to `x`, so for all intents and purposes we are implicitly saying that `test_approx(::NoTangent, x) = test_approx(zero(x), x)`.

Reply:
Rather than allowing it in general between `AbstractZero` and anything, which would allow between `NoTangent()` and `0.0` which seems bad.
would it be pragmatic to allow it only between `AbstractZero` and `NoTangent` ?


> Pirate test_approx(::NoTangent, x) in Zygote. (probably very bad)

Doing this in Zygote's tests though doesn't seem that awful.
(With appropriate comments)

Reply:
Oh, it is exactly the `NoTangent()` and `0.0` that we need, sorry that wasn't clear

Reply:
> We should think a bout the relationship between this and [JuliaDiff/ChainRulesCore.jl#363](https://github.com/JuliaDiff/ChainRulesCore.jl/pull/363)
> In particular that `frule_f` and `rrule_f`
> have basically (literally?) the same requirements as the `frule_via_ad` and `rrule_via_ad` from that PR.
> and so when using this for your AD you are probably having both defined.

Yeah that is a good point, I can't think of any differences on the top of my head. One consideration perhaps is how to deal with kwargs, since Zygote pretty much ignores them.

Reply:
Where do the `NoTangents` from Zygote come from?
`nothing` becomes `ZeroTangent`.
Which should allos the comparison

https://github.com/FluxML/Zygote.jl/blob/12f5c1d75eeaa8c7a818f2db7f8d082956c00cac/src/compiler/chainrules.jl#L62

Reply:
hmm Ok.

Reply:
Yeah, thanks, it's already changed in the code